django_setup_configuration/base.py

```diff
@@ -0,0 +1,174 @@
+from dataclasses import dataclass
+from typing import Iterator, Mapping, Sequence, Type
+
+from django.conf import settings
+from django.contrib.postgres.fields import ArrayField
+from django.db import models
+from django.db.models.fields import NOT_PROVIDED
+from django.db.models.fields.json import JSONField
+from django.db.models.fields.related import OneToOneField
+from django.utils.module_loading import import_string
+
+from .constants import basic_field_descriptions
+from .exceptions import ImproperlyConfigured
+
+
+@dataclass(frozen=True, slots=True)
+class ConfigField:
+    name: str
+    verbose_name: str
+    description: str
+    default_value: str
+    field_description: str
+
+
+class ConfigSettingsModel:
+    models: list[Type[models.Model]]
+    display_name: str
+    namespace: str
+    excluded_fields = ["id"]
+
+    def __init__(self, *args, **kwargs):
+        self.config_fields: list = []
+
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+        self.update_field_descriptions()
+
+        if not self.models:
+            return
+
+        for model in self.models:
+            self.create_config_fields(
+                exclude=self.excluded_fields,
+                model=model,
+            )
+
+    def get_setting_name(self, field: ConfigField) -> str:
+        return f"{self.namespace}_" + field.name.upper()
+
+    @staticmethod
+    def get_default_value(field: models.Field) -> str:
+        default = field.default
+
+        if default is NOT_PROVIDED:
+            return "No default"
+
+        # needed to make `generate_config_docs` idempotent
+        # because UUID's are randomly generated
+        if isinstance(field, models.UUIDField):
+            return "random UUID string"
+
+        # if default is a function, call the function to retrieve the value;
+        # we don't immediately return because we need to check the type first
+        # and cast to another type if necessary (e.g. list is unhashable)
+        if callable(default):
+            default = default()
+
+        if isinstance(default, Mapping):
+            return str(default)
+
+        # check for field type as well to avoid splitting values from CharField
+        if isinstance(field, (JSONField, ArrayField)) and isinstance(default, Sequence):
+            try:
+                return ", ".join(str(item) for item in default)
+            except TypeError:
+                return str(default)
+
+        return default
+
+    @staticmethod
+    def update_field_descriptions() -> None:
+        """
+        Add custom fields + descriptions defined in settings to
+        `basic_field_descriptions`
+        """
+        custom_fields = getattr(settings, "DJANGO_SETUP_CONFIG_CUSTOM_FIELDS", None)
```

Comment:
What is the shape of the setting and what does it do? It's never used but this line of code.  I think this class was designed to specify config step settings and related help texts, so it's the best place for custom help texts if they are needed.